File: src/services/ingestion/extractors/url-extractor.ts
import type { Document } from "../../../types/models.js";
import type { Extractor } from "./base.js";

/** Extractor for URL documents using fetch + basic HTML-to-text cleanup. */
export class UrlExtractor implements Extractor {
  /**
   * Fetches the URL and extracts readable text.
   * @param doc Source document.
   */
  public async extract(doc: Document): Promise<string> {
    const url = doc.sourceUrl ?? doc.rawContent;
    if (!url) {
      throw new Error("URL extractor requires sourceUrl or rawContent containing a URL");
    }

    const response = await fetch(url, {
      headers: {
        "user-agent": "Supermemento/2.0 (+https://supermemento.local)"
      }
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch URL (${response.status}): ${url}`);
    }

    const html = await response.text();
    return this.htmlToText(html);
  }

  private htmlToText(html: string): string {
    const withoutScripts = html
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, " ")
      .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, " ")
      .replace(/<noscript\b[^<]*(?:(?!<\/noscript>)<[^<]*)*<\/noscript>/gi, " ");

    const decoded = withoutScripts
      .replace(/<\s*br\s*\/?>/gi, "\n")
      .replace(/<\s*\/\s*(p|div|h1|h2|h3|h4|h5|h6|li|section|article)\s*>/gi, "\n")
      .replace(/<[^>]+>/g, " ")
      .replace(/&nbsp;/g, " ")
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">")
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'");

    return decoded
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
      .join("\n");
  }
}

File: src/services/ingestion/extractors/url_extractor.py
import re

from src.types.models import Document

from .base import Extractor


class UrlExtractor(Extractor):
    """Extractor for URL documents using fetch + basic HTML-to-text cleanup."""

    async def extract(self, doc: Document) -> str:
        """Fetches the URL and extracts readable text.
        
        Args:
            doc: Source document.
            
        Returns:
            Extracted text from HTML.
            
        Raises:
            ValueError: If URL is missing or fetch fails.
        """
        import aiohttp
        
        url = doc.source_url or doc.raw_content
        if not url:
            raise ValueError(
                "URL extractor requires sourceUrl or rawContent containing a URL"
            )

        async with aiohttp.ClientSession() as session:
            async with session.get(
                url,
                headers={
                    "User-Agent": "Supermemento/2.0 (+https://supermemento.local)"
                }
            ) as response:
                if response.status != 200:
                    raise ValueError(
                        f"Failed to fetch URL ({response.status}): {url}"
                    )
                html = await response.text()
                return self._html_to_text(html)

    def _html_to_text(self, html: str) -> str:
        """Convert HTML to plain text."""
        # Remove scripts and styles
        without_scripts = re.sub(
            r'<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>',
            ' ',
            html,
            flags=re.IGNORECASE | re.DOTALL
        )
        without_styles = re.sub(
            r'<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>',
            ' ',
            without_scripts,
            flags=re.IGNORECASE | re.DOTALL
        )
        without_noscript = re.sub(
            r'<noscript\b[^<]*(?:(?!<\/noscript>)<[^<]*)*<\/noscript>',
            ' ',
            without_styles,
            flags=re.IGNORECASE | re.DOTALL
        )

        # Replace common tags with newlines
        decoded = without_noscript
        decoded = re.sub(r'<\s*br\s*\/?>', '\n', decoded, flags=re.IGNORECASE)
        decoded = re.sub(
            r'<\s*\/\s*(?:p|div|h[1-6]|li|section|article)\s*>',
            '\n',
            decoded,
            flags=re.IGNORECASE
        )
        
        # Remove remaining tags
        decoded = re.sub(r'<[^>]+>', ' ', decoded)
        
        # Decode HTML entities
        decoded = decoded.replace('&nbsp;', ' ')
        decoded = decoded.replace('&amp;', '&')
        decoded = decoded.replace('&lt;', '<')
        decoded = decoded.replace('&gt;', '>')
        decoded = decoded.replace('&quot;', '"')
        decoded = decoded.replace('&#39;', "'")

        # Clean up whitespace
        lines = [
            line.strip() 
            for line in decoded.split('\n') 
            if line.strip()
        ]
        return '\n'.join(lines)
